k8srm-prototype/pkg/api/capacity_types.go

```diff
@@ -38,19 +39,91 @@ type DevicePoolSpec struct {
 	// Attributes contains device attributes that are common to all devices
 	// in the pool.
 	// +optional
+	// +listType=atomic
 	Attributes []Attribute `json:"attributes,omitempty"`
 
-	// DeviceCount contains the total number of devices in the pool.
+	// SharedResources are pooled resources that are shared by all devices
+	// in the pool. This is typically used when representing a
+	// partitionable device, and need not be populated otherwise.
+	//
+	// +optional
+	// +listType=atomic
+	SharedResources []ResourceCapacity `json:"sharedResources,omitempty"`
+
+	// Devices contains the individual devices in the pool.
+	//
 	// +required
-	DeviceCount int `json:"count,omitempty"`
+	// +listType=atomic
+	Devices []Device `json:"devices,omitempty"`
 }
 
 // DevicePoolStatus contains the state of the pool as last reported by the
 // driver. Note that this will not include the allocations that have been made
 // by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
 // to make future scheduling decisions.
 type DevicePoolStatus struct {
-	AvailableDevices int `json:"availableDevices,omitempty"`
+	DeviceStatuses []AllocatedDevice `json:"deviceStatuses,omitempty"`
+}
+
+// AllocatedDevice represents a device that has been allocated from the pool.
+type AllocatedDevice struct {
+	Name string
+	// Conditions contains the latest observation of the device's state.
+	Conditions []metav1.Condition `json:"conditions"`
+
+	// ClaimUIDs contains the UIDs of the claims to which this device
+	// is allocated.
+	ClaimUIDs []types.UID
+}
+
+// Device is used to track individual devices in a pool.
+type Device struct {
+	// Name is a driver-specific identifier for the device.
+	// +required
+	Name string `json:"name"`
+
+	// Attributes contain additional metadata that can be used in
+	// constraints. If an attribute name overlaps with the pool attribute,
+	// the device attribute takes precedence.
+	//
+	// +optional
+	Attributes []Attribute `json:"attributes,omitempty"`
+
+	// ConsumesSharedResources contains the pooled resources that are
+	// consumed when this device is allocated.
+	//
+	// +optional
+	ConsumesSharedResources map[string]resource.Quantity `json:"consumesSharedResources,omitempty"`
+
+	// ProvidesClaimResources allows the definition of per-device resources
+	// that can be allocated in a manner similar to standard Kubernetes
+	// resources.
+	//
+	// +optional
+	ProvidesClaimResources []ResourceCapacity `json:"providesClaimResources,omitempty"`
```

Comment:
I think `allocatable` was right for this`?

Reply:
Well, I was trying to make it even more explicit. How about `AllocatableClaimResources`? I think just `Allocatable` is not so clear that it means from the claim? Trying to make a very clear distinction between the shared stuff (internal to the capacity model) and the user-facing stuff.